core/arch/arm/kernel/thread_spmc.c

```diff
@@ -1843,27 +1848,51 @@ void thread_spmc_relinquish(uint64_t cookie)
 		EMSG("Failed to relinquish cookie %#"PRIx64, cookie);
 }
 
-static int set_pages(struct ffa_address_range *regions,
-		     unsigned int num_regions, unsigned int num_pages,
-		     struct mobj_ffa *mf)
+static int add_pages_at(struct ffa_address_range *regions,
+			unsigned int num_regions, unsigned int *idx,
+			struct mobj_ffa *mf)
 {
 	unsigned int n = 0;
-	unsigned int idx = 0;
 
 	for (n = 0; n < num_regions; n++) {
 		unsigned int page_count = READ_ONCE(regions[n].page_count);
 		uint64_t addr = READ_ONCE(regions[n].address);
 
-		if (mobj_ffa_add_pages_at(mf, &idx, addr, page_count))
+		if (mobj_ffa_add_pages_at(mf, idx, addr, page_count))
 			return FFA_INVALID_PARAMETERS;
 	}
 
-	if (idx != num_pages)
-		return FFA_INVALID_PARAMETERS;
-
 	return 0;
 }
 
+static void *spmc_frag_retrieve(uint64_t cookie,
+				uint32_t next_frag_offset,
+				uint32_t *fragment_length)
+{
+	struct thread_smc_args args = {
+		.a0 = FFA_MEM_FRAG_RX,
+		.a1 = (uint32_t)(cookie),        /* handle high addr */
+		.a2 = (uint32_t)(cookie >> 32),  /* handle low addr */
+		.a3 = next_frag_offset,          /* offset */
+		.a4 = 0,			 /* MBZ */
```

Comment:
This assignment can be skipped. But a comment would be nice, saying something like:
```
/* w4 MBZ since we're a virtual instance */
```

Reply:
> 

Do you mean

Reply:
No, just the comment.

Reply:
Done